update: ksp-incremental.md (#5583)
JetBrains/kotlin-web-site: docs/topics/ksp/ksp-incremental.md

[//]: # (title: Incremental processing)

KSP supports incremental processing: KSP reprocesses a file only when one or more of its dependencies change. This 
avoids unnecessary reprocessing and therefore reduces compilation time.

Incremental processing is enabled by default. To disable it, add this line to your `gradle.properties` file:

```
ksp.incremental=false
```

## Dirty files

A file is considered _dirty_ if it is either directly modified by the user or indirectly affected by changes in other

> dirty files. A dirty source requires reprocessing.

Reviewer: I think "requires reprocessing" is a part of definition of a dirty source rather than a consequence? This feels closer, but feel free to correct me: "A file is considered _dirty_ (needing to be reprocessed) if..."

To determine which sources are dirty, KSP relies on processors, which associate generated outputs with their 
corresponding input sources. KSP uses these associations to identify the sources that must be reprocessed when a change 
occurs.

KSP requires only a minimal set of root sources. Processors use these sources as entry points to navigate the code structure.

When a processor generates an output, it must associate that output with the sources of the corresponding `KSNode`. This 
association is required if the `KSNode` is obtained from any of the following methods:

* `Resolver.getAllFiles()`
* `Resolver.getSymbolsWithAnnotation()`
* `Resolver.getClassDeclarationByName()`
* `Resolver.getDeclarationsFromPackage()`

> > To learn more about the `KSNode` API, see the diagram in [How KSP models Kotlin code](https://kotlinlang.org/docs/ksp-additional-details.html).

Reviewer: I would introduce this link before KSNode: those who read it already would at least understand where we are, those who haven't can, like me, panic at the sight of `KSNode` and start googling before the paragraph ends.

Also this link should not be absolute, right? If it's a topic in the same repo.

Author: In the end, the dev and I rewrote this section a bit for clarity and removed the mention of KSNode altogether.

> 
> {style="tip"}

### Aggregating and isolating outputs

KSP classifies generated outputs into two types: aggregating and isolating.

> Unlike Gradle annotation processing, KSP applies the categorization to individual outputs
> rather than entire processors.
>
{style=”note”}

<deflist collapsible="true">
<def title="Aggregating">

Aggregating outputs can be affected by changes in any source file, except for removals that do not impact other files.

Any input change triggers a rebuild of all aggregating outputs and reprocessing of all corresponding registered, new,
or modified source files.

For example, an output that collects all symbols with a particular annotation is aggregating.

</def>
<def title="Isolating">

Isolating outputs depend only on their specified sources.

Changes to other sources do not affect the output. Multiple source files can be associated with a single output

For example, a generated class that is dedicated to an interface it implements is isolating.

</def>
</deflist>

### Dirtiness propagation

> KSP propagates dirtiness in two steps, which are repeated until there are no dirty files:

Reviewer: Suddenly I'm thinking that "propagating dirtiness until there are no dirty files" doesn't make sense. Does it? Dirty files are going to be reprocessed, that's how we get rid of dirty files; until then we just need to know which ones are dirty, and for that we monitor changes + the propagation mechanism?

Author: Yes, you are right. The two things can trigger each other, though, and there can be some back and forth before it's over. That's where the idea of repetition "until there are no new dirty files" came from. More accurate would be to say "until no new files are marked as dirty". I'll rephrase.

[//]: # (title: Incremental processing)

KSP supports incremental processing: KSP reprocesses a file only when one or more of its dependencies change. This
avoids unnecessary reprocessing and therefore reduces compilation time.

Incremental processing is enabled by default. Disable it when troubleshooting or forcing a full rebuild. To disable it, 

> add this line to your `gradle.properties` file:

Reviewer 2: What do you think about the following option, does it work?

> Incremental processing is enabled by default. You can disable it when troubleshooting or when you need to force a full rebuild. To disable it, add the following line to your gradle.properties file:

Author: Yes, better. On it!

[//]: # (title: Incremental processing)

KSP supports incremental processing: KSP reprocesses a file only when one or more of its dependencies change. This
avoids unnecessary reprocessing and therefore reduces compilation time.

Incremental processing is enabled by default. To disable it, add this line to your `gradle.properties` file:

```
ksp.incremental=false
```

> You might want to disable incremental processing when troubleshooting or to force a full rebuild.

Reviewer 2: I believe that this line should be somewhere (mb even rephrased) before the instruction, how to disable the IC

## Dirty files

> A file is considered _dirty_ (needing to be reprocessed) if it is either directly modified by the user or indirectly 

Reviewer 2: A "user" here is the developer who reads this doc? 
In this case I'd like to make it more obvious.

Reviewer 2: BTW, we avoid saying "User" when we mean "A developer"

root sources and their tracked dependencies to determine when the output needs to be regenerated.

> Use the `CodeGenerator` interface to create output files and associate inputs with outputs. For more information, 
> see [`CodeGenerator.kt` in the source code](https://github.com/google/ksp/blob/main/api/src/main/kotlin/com/google/devtools/ksp/processing/CodeGenerator.kt).
>
{style="tip"}

### Aggregating and isolating outputs

KSP classifies generated outputs into two types: aggregating and isolating.

> Unlike Gradle annotation processing, KSP applies the categorization to individual outputs
> rather than entire processors.
>
{style="note"}

<deflist collapsible="true">
<def title="Aggregating">

Aggregating outputs can be affected by changes in any source file, except for removals that do not impact other files.

Any input change triggers a rebuild of all aggregating outputs and reprocessing of all corresponding registered, new,
or modified source files.

For example, an output that collects all symbols with a particular annotation is aggregating.

</def>
<def title="Isolating">

Isolating outputs depend only on their specified sources.

> Changes to other sources do not affect the output. Multiple source files can be associated with a single output

Reviewer 3: missing full stop

[//]: # (title: Incremental processing)

> KSP supports incremental processing. KSP reprocesses a file only when one or more of its dependencies change. This 

Reviewer: ```suggestion
KSP supports incremental processing: KSP reprocesses a file only when one or more of its dependencies change. This 
```

Since these sentences are basically the same statement, I have the urge to connect them somehow.

> reduces compilation time by avoiding unnecessary reprocessing.

Reviewer: "This avoids unnecessary reprocessing and therefore reduces compilation time" sounds a bit clearer to me; verbs are better than gerunds and especially gerund chains.

> Incremental processing is enabled by default. To disable it, add this line to your `gradle.properties` file:

Reviewer: Is there a common reason to disable it? I'd add it in parentheses here, to anchor the use case.

```
ksp.incremental=false
```

> For general information, see Wikipedia's article on [incremental computing](https://en.wikipedia.org/wiki/Incremental_computing).
>

> {style=”tip”}

Reviewer: How useful is this link here? It is general information indeed, but I'm afraid it might be too general :)

## Dirty files

> A source that requires reprocessing is considered dirty. To determine which sources are dirty, KSP relies on processors 

Reviewer: ```suggestion
A source that requires reprocessing is considered _dirty_. To determine which sources are dirty, KSP relies on processors 
```

> to associate generated outputs with their corresponding input sources. KSP uses these associations to identify the 

Reviewer: "To do X, KSP relies on processors to do Y" can be confusing — would it be correct to say "...KSP relies on processors, which associate..."?

> sources that must be reprocessed when changes occur.

Reviewer: ```suggestion
sources that must be reprocessed when a change occurs.
```

> KSP requires only a minimal set of root sources. Processors use these sources as entry points to navigate the code structure.

Reviewer: The concept of root sources and the mention of `KSNode` below doesn't paint a picture for me, although I'm sure it should. The general idea of "processors match outputs and sources → KSP uses this matching to find sources to be reprocessed" is clear; what role do root sources play here? What is `KSNode` and how is it related to root sources?

When a processor generates an output, it must associate that output with the sources of the corresponding `KSNode`. This 

> association is required if the `KSNode` is obtained from any of the following methods:

Reviewer: I feel that these are conflicting statements: either a processor must always associate an output with sources, or the association is only required if a condition is met. Am I missing something?

* `Resolver.getDeclarationsFromPackage()`

## Aggregating and isolating outputs

KSP classifies generated outputs into two types: aggregating and isolating.

| **Output type**   | **Description**                                                                                   | **Behavior**                                                                                                                                   | **Example**                                                       |
|-------------------|---------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------------------------------|-------------------------------------------------------------------|
| Aggregating       | Can be affected by changes in any source file, except for removals that do not impact other files | Any input change triggers a rebuild of all aggregating outputs and reprocessing of all corresponding registered, new, or modified source files | An output that collects all symbols with a particular annotation  |

> | Isolating         | Depends only on its specified sources                                                             | Changes to other sources do not affect the output. Multiple source files can be associated with a single output                                | A generated class that is dedicated to an interface it implements |

Reviewer: In the current kotlinlang.org frontend, the table is almost unusable — maybe the rows should become columns, or columns should have fixed width. That's one of the reasons we generally use `<table>`, not Markdown tables.

> Unlike Gradle annotation processing, KSP applies the categorization to individual outputs
> rather than entire processors.
>

> {style=”note”}

Reviewer: Maybe this note can be after the first paragraph here? The assumption that the categorization works the same should probably be dealt with as soon as it appears.

## Implementation

Dependencies are determined by the many-to-many relationship between input and output files.

This is how KSP determines which files need to be reprocessed:

* If an input file is changed, it will always be reprocessed.

    **Why?** If an input is changed, new information can be introduced. Processors need to run again with the input.

* If an input file is changed and is associated with an output, then all other input files associated with the same 
output will also be reprocessed. This happens repeatedly until there is no new dirty file.

    **Why?** An output is made out of a set of inputs. Processors may need all the inputs to regenerate the output.

* If an unchanged input file isn't associated with any aggregating outputs, it won't be reprocessed.

    **Why?** This file can't affect any outputs because it is unchanged and isn't associated with an aggregating output. 
It won't be reprocessed unless one of the above rules applies.

## Dirtiness propagation

A file is considered dirty if it is either directly modified by the user or indirectly affected by changes in other dirty files.

> KSP propagates dirtiness in two steps:

Reviewer: If these are sequential, the following list should probably be numbered.

* Propagation by **resolution tracing**: Type resolution is the only way to traverse from one file to another. When a 
processor resolves a type reference (explicitly or implicitly), KSP considers dependencies between the file containing 
the reference and any file that defines a symbol affecting that resolution. As a result, a change in a resolved symbol 
may mark the referencing file as dirty.

* Propagation by **input-output correspondence**: If a source file is changed or affected, all other source files that 
share generated outputs with it are also marked as affected. This groups related files into equivalence classes based on 
shared outputs.

> Note that both of these steps will be repeated until there are no dirty files.

Reviewer: I'd add this to the intro paragraph: "KSP propagates dirtiness in two steps, which are repeated until there are no dirty files."

> Unlike Gradle annotation processing, KSP applies the categorization to individual outputs
> rather than entire processors.
>
{style=”note”}

## Implementation

Dependencies are determined by the many-to-many relationship between input and output files.

This is how KSP determines which files need to be reprocessed:

* If an input file is changed, it will always be reprocessed.

    **Why?** If an input is changed, new information can be introduced. Processors need to run again with the input.

* If an input file is changed and is associated with an output, then all other input files associated with the same 
output will also be reprocessed. This happens repeatedly until there is no new dirty file.

    **Why?** An output is made out of a set of inputs. Processors may need all the inputs to regenerate the output.

* If an unchanged input file isn't associated with any aggregating outputs, it won't be reprocessed.

>     **Why?** If an output file is categorized as isolating, the processor is sure that the information comes from a 

Reviewer 4: It seems a bit unclear to me how the 'why' part answers the statement. Maybe it's clearer to say something along the lines "Since it's not associated with an output and it's unchanged (meaning not new nor modified), it can not contribute to any output change."
I see that the reason is borrowed from the original document and the new version is not incorrect. Maybe you could also write it as "_If an output file is categorized as not aggregating, it must be isolating, so the processor_ is sure that information comes from a ..."

Additionally, it could be a bit more precise to update the statement to "If an unchanged input file ..., it won't be reprocessed _unless one of the rules above apply_".

specific set of input files. If those files don’t change, neither will the output file.

## Dirtiness propagation

A file is considered dirty if it is either directly modified by the user or indirectly affected by changes in other dirty files.
KSP propagates dirtiness in two steps:

* Propagation by **resolution tracing**: Type resolution is the only way to traverse from one file to another. When a 
processor resolves a type reference (explicitly or implicitly), KSP considers dependencies between the file containing 
the reference and any file that defines a symbol affecting that resolution. As a result, a change in a resolved symbol 
may mark the referencing file as dirty.

* Propagation by **input-output correspondence**: If a source file is changed or affected, all other source files that 
share generated outputs with it are also marked as affected. This groups related files into equivalence classes based on 
shared outputs.

Note that both of these steps will be repeated until there are no dirty files.

## Example 1

A processor generates `outputForA` after reading class A in `A.kt` and class B in `B.kt`, where A extends B.

> The processor gets A by `Resolver.getSymbolsWithAnnotation` and B by `KSClassDeclaration.superTypes` from A.

Reviewer 4: This is just a suggestion (for both examples): The files could be listed in a `tree` view, so readers will recognize the file tree view. Additionally, the code that the processors execute could be formatted in an ordered list so it's clear what happens when. E.g.,

```
.
├── A.kt
└── B.kt
```
The processor generates `outputForA` by:

1. Obtains `A` by calling `Resolver.getSymbolsWithAnnotation`
2. Obtains `B` by calling `KSClassDeclaration.superTypes` on `A`.

Reviewer 4: For the second example, the order in which files are read is a bit unclear so again an ordered list there might also help.

Because the inclusion of B is due to A, `B.kt` doesn't need to be specified in dependencies for `outputForA`. You can 

> still specify `B.kt` in this case, but it is unnecessary.

Reviewer 4: Maybe this can be clarified a bit by saying something like "... since `B` is obtained from `A`, KSP associates `B` as a dependency of `A` via resolution tracing."

### Dirtiness propagation

> KSP propagates dirtiness in two ways:

Reviewer: I'm allergic to hardcoding the number of things in a list, so I would just write "in the following ways".

> 1. by **resolution tracing**: Type resolution is the only way to traverse from one file to another. When a

Reviewer: Lower case looks strange here, let's capitalize the beginning of the sentence.

* If an input file is changed, it will always be reprocessed.

> **Why?** If an input is changed, new information can be introduced. Processors need to run again with the input.

Reviewer: Not sure why the indent was removed — now a comment to a list item is on a higher level than the thing it's commenting on, and visually it's hard to track their relationship.

}
```

To generate `outputForA`, the processor:

1. gets A by calling `Resolver.getSymbolsWithAnnotation`.

2. gets B by calling `KSClassDeclaration.superTypes` on A.

KSP tracks this relationship through resolution tracing and automatically records B as a dependency of A. Therefore, 
the processor doesn't need to explicitly declare B.kt as a dependency of outputForA.

## Example 2

```none
.
├── src
│   ├── sourceA.kt
│   └── sourceB.kt
└── generated
    ├── outputA
    └── outputB
```

A processor:

> 1. reads `sourceB`.

Reviewer: Is there a meaning behind reading B before A? If not, doing the default order is going to be less prone to questions like that )

2. generates `outputB`.

3. reads `sourceA`.

4. generates `outputA`.

> If `sourceA` changes:

Reviewer: ```suggestion
When `sourceA` changes:
```

* If `outputB` is aggregating, KSP reprocesses both `sourceA` and `sourceB`.

* If `outputB` is isolating, KSP reprocesses `sourceA` only.

If `sourceC` is added:

* If `outputB` is aggregating, KSP reprocesses `sourceC` and `sourceB`.

* If `outputB` is isolating, KSP reprocesses only `sourceC`.

If either `sourceA` or `sourceB` is removed, KSP doesn't need to reprocess any files.

## Reporting bugs

> If you encounter any error that only occurs in incremental mode, report it by creating an issue in GitHub. Follow the 

Reviewer: There should probably be a link to the repo where the issue should be created.

> If you encounter any error that only occurs in incremental mode, report it by creating an issue in GitHub. Follow the 

Reviewer: ```suggestion
If you encounter any error that only occurs in incremental mode, report it by creating an issue in GitHub.
```

> following steps to retrieve the relevant logs:

Reviewer: ```suggestion
Attach the relevant logs to the issue:
```

}
```

To generate `outputForA`, the processor:

1. gets A by calling `Resolver.getSymbolsWithAnnotation`.

2. gets B by calling `KSClassDeclaration.superTypes` on A.

KSP tracks this relationship through resolution tracing and automatically records B as a dependency of A. Therefore, 
the processor doesn't need to explicitly declare B.kt as a dependency of outputForA.

## Example 2

```none
.
├── src
│   ├── sourceA.kt
│   └── sourceB.kt
└── generated
    ├── outputA

>     └── outputB

Reviewer 4: nice :)

```

> A processor generates `outputA` after reading `sourceA` and `outputB` after reading `sourceB`.

Reviewer 4: I still find this sentence confusing. Could we make an ordered list here that shows exactly what happens when? E.g.,

1. Reads sourceB
2. Generates outputB
3. Reads sourceA
4. Generates outputA

If `sourceA` changes:

* If `outputB` is aggregating, KSP reprocesses both `sourceA` and `sourceB`.

> * If `outputB` is isolating, KSP reprocesses `sourceA`only.

Reviewer 4: Missing space at the end :)

## Example 2

A processor generates `outputA` after reading `sourceA` and `outputB` after reading `sourceB`.

When `sourceA` is changed:

* If outputB is aggregating, both sourceA and sourceB are reprocessed.

> * If outputB is isolating, only sourceA is reprocessed.

Reviewer 4: Nit: should `sourceA`, `sourceB`, and `outputB` be wrapped in backticks?

Author: Yes, of course. Missed it!

       val dependencies = Dependencies(aggregating = true, declA.containingFile!!)
       // outputForA.kt
       val outputName = "outputFor${declA.simpleName.asString()}"
       // outputForA depends on A.kt and B.kt
       val output = codeGenerator.createNewFile(dependencies, "com.example", outputName, "kt")
       output.write("// $declA : $declB\n".toByteArray())
       output.close()
   }
   // ...
}
```

To generate `outputForA`, the processor:

1. gets A by calling `Resolver.getSymbolsWithAnnotation`.

2. gets B by calling `KSClassDeclaration.superTypes` on A.

KSP tracks this relationship through resolution tracing and automatically records `B` as a dependency of `A`. Therefore,
you don't need to explicitly declare `B.kt` as a dependency of `outputForA`.

## Reporting bugs

If you encounter any error that only occurs in incremental mode, report it by creating an issue in
[the GitHub repository](https://github.com/google/ksp/issues). Attach the relevant logs to the issue:

1. Enable logs that dump the dirty set according to dependencies and outputs. To do this, add this line to
   `gradle.properties`:

  ```
  ksp.incremental.log=true
   ```

> 2. Perform a clean build. This build produces two log files in the build output directory with a `.log` file extension:

Reviewer 3: if the user is reporting a bug related to incrementality - performing a clean build will actually fix the issue.

Reviewer 3: Maybe we can say 

2. Perform a clean build to initialize the caches. 
3. Make a change to a source file that triggers the issue, and run the build again. 
4. Collect the log files produced in the build output directory

Reviewer 4: Also, I'm not sure the last two logs are generated. There is no code path at least that reaches these names:
`build/kspCaches/<source set>/logs/kspDirtySetByDeps.log`
`build/kspCaches/<source set>/logs/kspDirtySetByOutputs.log`.
The `kspDirtySet.log` and `kspSourceToOutputs.log` are generated though.
For error reporting, they should have the logs from the successful build as well as the failed build so we can check what the diff is. So the idea is that they should copy the generated logs from the successful build somewhere else first, then perform the change and run the incremental build which regenerates the logs which can then be copied somewhere else. That should in total yield 4 files.
Just wanted to point it out and clear up the mismatch between the code and the documentation :)

Author: Could you tell me what the step by step would be then? Something like this?

1. Perform a clean build (successful)
2. save the logs somewhere
3. modify a source file, triggering the issue
4. run the build again (failed)
5. collect the logs

Author: And this part should not be there at all:

```
You can then run successive incremental builds, which will generate two additional log files:

* `build/kspCaches/<source set>/logs/kspDirtySetByDeps.log`
* `build/kspCaches/<source set>/logs/kspDirtySetByOutputs.log`
```

Correct?

Reviewer 4: > And this part should not be there at all:
> 
> ```
> You can then run successive incremental builds, which will generate two additional log files:
> 
> * `build/kspCaches/<source set>/logs/kspDirtySetByDeps.log`
> * `build/kspCaches/<source set>/logs/kspDirtySetByOutputs.log`
> ```
> 
> Correct?

Yes, exactly.

Reviewer 4: > Could you tell me what the step by step would be then? Something like this?
> 
> 1. Perform a clean build (successful)
> 2. save the logs somewhere
> 3. modify a source file, triggering the issue
> 4. run the build again (failed)
> 5. collect the logs

Yes that's correct :)

2. gets B by calling `KSClassDeclaration.superTypes` on A.

KSP tracks this relationship through resolution tracing and automatically records `B` as a dependency of `A`. Therefore,
you don't need to explicitly declare `B.kt` as a dependency of `outputForA`.

## Reporting bugs

If you encounter any error that occurs only when incremental processing is enabled, report it by creating an issue in
[the GitHub repository](https://github.com/google/ksp/issues). Attach the relevant logs to the issue:

1. Enable incremental processing logs by adding the following line to `gradle.properties`:

  ```
  ksp.incremental.log=true

>    ```

Reviewer: The indentation is a little mismatched here.
